pallets/pallet-lottery/src/staking.rs

```diff
@@ -0,0 +1,495 @@
+
+        // first concern: If we fell out of the active set on one or more collators, we need to get back into it
+        let active_collators = pallet_parachain_staking::Pallet::<T>::selected_candidates()
+            .into_iter()
+            .collect();
+        let collators_we_are_staked_with: BTreeSet<_> = StakedCollators::<T>::iter_keys().collect();
+        let active_collators_we_are_staked_with =
+            collators_we_are_staked_with.intersection(&active_collators);
+
+        let mut remaining_deposit = new_deposit;
+        for collator in active_collators_we_are_staked_with {
+            let staked = StakedCollators::<T>::get(collator.clone());
+            let info = pallet_parachain_staking::Pallet::<T>::candidate_info(collator.clone())
+                .expect("is active collator, therefore it has collator info. qed");
+            if staked < info.lowest_top_delegation_amount {
+                // TODO: Small optimization: sort collators ascending by missing amount so we get the largest amount of collators active before running out of funds
+                let deposit =
+                    remaining_deposit.saturating_sub(info.lowest_top_delegation_amount - staked);
+                deposits.push((collator.clone(), deposit));
+                remaining_deposit = remaining_deposit.saturating_sub(deposit);
+                if remaining_deposit.is_zero() {
+                    break;
+                }
+            }
+        }
+        // If we have any collators to re-activate, we distribute all tokens to those and call it a day
+        if !deposits.is_empty() {
+            if !remaining_deposit.is_zero() {
+                // distribute remaining tokens evenly
+                let deposit_per_collator =
+                    Percent::from_rational(1, deposits.len()).mul_ceil(remaining_deposit); // this overshoots the amount if there's a remainder
+                for deposit in &mut deposits {
+                    let add = remaining_deposit.saturating_sub(deposit_per_collator);
+                    deposit.1 += add;
+                    remaining_deposit -= add;
+                }
+            }
+            return deposits;
+        }
+
+        // second concern: We want to maximize staking APY earned, so we want to balance the staking pools with our deposits while conserving gas
+        // We only consider active collators for deposits
+        // TODO: Small optimization: Also consider points / pointsAwarded to not stake to collators missing blocks
+
+        let mut top_collator_accounts =
+            pallet_parachain_staking::Pallet::<T>::compute_top_candidates(); // XXX: This can select collators that are joined but not yet producing blocks
+        if top_collator_accounts.is_empty() {
+            // Use `SelectedCandidates` as fallback (should basically never happen)
+            log::warn!("Lottery falling back to selected candidates");
```

Comment:
maybe put a why this is a warning in the message

Reply:
done